lib/operations/aggregate.js

```diff
@@ -1,132 +1,122 @@
 'use strict';
 
-const AggregationCursor = require('../aggregation_cursor');
-const applyWriteConcern = require('../utils').applyWriteConcern;
-const decorateWithCollation = require('../utils').decorateWithCollation;
-const decorateWithReadConcern = require('../utils').decorateWithReadConcern;
-const handleCallback = require('../utils').handleCallback;
+const CommandOperationV2 = require('./command_v2');
 const MongoError = require('../core').MongoError;
-const resolveReadPreference = require('../utils').resolveReadPreference;
-const toError = require('../utils').toError;
+const maxWireVersion = require('../core/utils').maxWireVersion;
 const ReadPreference = require('../core').ReadPreference;
+const Aspect = require('./operation').Aspect;
+const defineAspects = require('./operation').defineAspects;
 
 const DB_AGGREGATE_COLLECTION = 1;
 const MIN_WIRE_VERSION_$OUT_READ_CONCERN_SUPPORT = 8;
 
-/**
- * Perform an aggregate operation. See Collection.prototype.aggregate or Db.prototype.aggregate for more information.
- *
- * @method
- * @param {Db} db A Db instance.
- * @param {Collection|string} coll A collection instance or the string '1', used for db.aggregate.
- * @param {object} [pipeline=[]] Array containing all the aggregation framework commands for the execution.
- * @param {object} [options] Optional settings. See Collection.prototype.aggregate or Db.prototype.aggregate for a list of options.
- * @param {Db~aggregationCallback|Collection~aggregationCallback} callback The command result callback
- */
-function aggregate(db, coll, pipeline, options, callback) {
-  const isDbAggregate = typeof coll === 'string';
-  const target = isDbAggregate ? db : coll;
-  const topology = target.s.topology;
-  let hasWriteStage = false;
-
-  if (typeof options.out === 'string') {
-    pipeline = pipeline.concat({ $out: options.out });
-    hasWriteStage = true;
-  } else if (pipeline.length > 0) {
-    const finalStage = pipeline[pipeline.length - 1];
-    if (finalStage.$out || finalStage.$merge) {
-      hasWriteStage = true;
+class AggregateOperation extends CommandOperationV2 {
+  constructor(parent, pipeline, options) {
+    // ensure we receive an unchanged raw response from the server (for cursor logic)
+    options.full = true;
+    super(parent, options);
+
+    this.target =
```

Comment:
I'm not sure if this is entirely reliable. While it was implied, I did not codify an explicit guarantee that a db's namespace would *not* have a collection.

Reply:
how do you imagine this would fail? If we don't have a collection, we have no way to build the aggregation. I think it's safe to assume in those cases you're running against the database, and to rely on a server error

Reply:
It's mostly just that I was toying around with the idea of a database namespace defaulting to `db.$cmd` to simplify executing commands on a Db. I think it's fine for now, I just wanna make sure we remember it.